### PR TITLE
Don't bind RPC.Client to fixed routing key

### PR DESCRIPTION
**Pull request by take-five**
After using RPC.Client in real project, I've found out that
binding RPC Client to a single routing key leads to bad design. E.g.
we have to spawn several similar processes to communicate with
several services whilst 1 process could do the same.

This commit allows to have one RPC.Client process for multiple services.

For example:
```
defmodule MyApp.RPC.Client do
  use Freddy.RPC.Client

  def start_link(conn), do: Freddy.RPC.Client.start_link(__MODULE__, conn, [], nil, name: Client)
end

defmodule MyApp.SitesService do
  def get_sites(), do: MyApp.RPC.Client.request(Client, "SitesQueue", %{type: "get_sites"})
end

defmodule MyApp.UsersService do
  def get_users(), do: MyApp.RPC.Client.request(Client, "SitesQueue", %{type: "get_useres"})
end
```

One RPC.Client process can handle multiple requests at the same time.

Also, interface of `before_request`, `on_response` and `on_timeout` callbacks
has been changed. Now these callbacks accept `Freddy.RPC.Request` data structure
as a first argument. This structure holds all necessary request information: payload,
routing key, options etc. In addition, callback `on_return` has been added which
mirrors `on_return` callbacks from `hare` and follows the same semantics as the
callbacks, mentioned above.

### CODE REVIEW

File: lib/freddy/rpc/client.ex [hunk 1]
```diff
@@ -376,34 +377,71 @@ defmodule Freddy.RPC.Client do
...
+    encoded_payload = Poison.encode!(request.payload)
+    routing_key = request.routing_key
+    freddy_options = [mandatory: true,
+                      content_type: "application/json",
+                      expiration: to_string(state.timeout),
+                      type: "request"]
+    options = Keyword.merge(request.options, freddy_options)
+    new_state = state |> State.update(new_given) |> State.push_waiting(request)
+
+    Logger.debug fn -> "Sending request to #{routing_key}: #{encoded_payload}" end
```
Comment (indrekj) [inline lib/freddy/rpc/client.ex]: when should metadata be used and when should it be inlined?

Comment (take-five) [inline lib/freddy/rpc/client.ex]: > when should metadata be used and when should it be inlined?
 
this is application-specific logic, this library shouldn't know about it
Actually, I'm inclined to remove logging from Freddy completely and let applications handle it

File: lib/freddy/rpc/client.ex [hunk 2]
```diff
@@ -68,25 +69,22 @@ defmodule Freddy.RPC.Client do
...
+  It receives as argument the RPC request structure which contains the message
+  payload, the routing key and the options for that publication, and the
+  internal client state.
 
   Returning `{:ok, state}` will cause the request to be performed with no
   modification, block the client until the response is received, and enter
   the main loop with the given state.
 
-  Returnung `{:ok, meta, state}` will do the same as `{:ok, state}`, but
+  Returnung `{:ok, request, state}` will do the same as `{:ok, state}`, but
```
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: typo `Returnung`

File: lib/freddy/rpc/client.ex [hunk 3]
```diff
@@ -242,23 +270,24 @@ defmodule Freddy.RPC.Client do
         do: :ok
 
       defoverridable [init: 1, terminate: 2,
-                      handle_ready: 2, before_request: 3,
-                      on_response: 3, on_timeout: 2,
+                      handle_ready: 2, before_request: 2,
+                      on_response: 3, on_timeout: 2, on_return: 2,
```
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: How is this change related to the routing key?

File: lib/freddy/rpc/client.ex [hunk 4]
```diff
@@ -376,34 +377,71 @@ defmodule Freddy.RPC.Client do
     end
   end
 
+  defp send_request(request, new_given, state) do
+    encoded_payload = Poison.encode!(request.payload)
```
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: Why are we relying in exceptions instead of success/failure return values? Should `send_request` be changed to `send_request!` by the same convention?

File: lib/freddy/rpc/client.ex [hunk 5]
```diff
@@ -376,34 +377,71 @@ defmodule Freddy.RPC.Client do
...
+    routing_key = request.routing_key
```
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: Could this be pattern matched from the function arguments? Why prefer one over the other?

File: lib/freddy/rpc/client.ex [hunk 6]
```diff
@@ -376,34 +377,71 @@ defmodule Freddy.RPC.Client do
...
+    end
   end
 
   @doc false
   def on_timeout(from, %{mod: mod, given: given} = state) do
-    {meta, new_state} = State.pop_waiting(state, from)
+    {request, new_state} = State.pop_waiting(state, from)
 
-    log_timeout(meta, new_state)
+    if request do
```
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: wouldn't using `{:ok, request, new_state}` and `:not_found` be nicer here than using `nil`?

File: lib/freddy/rpc/client.ex [hunk 7]
```diff
@@ -488,19 +532,17 @@ defmodule Freddy.RPC.Client do
...
     response
   end
 
-  defp log_timeout(meta, %{routing_key: queue}),
-    do: Logger.error fn -> ["RPC call to ", queue, " timed out after #{calculate_elapsed(meta)} ms"] end
-
-  defp log_return(_meta, %{routing_key: queue}),
-    do: Logger.error fn -> ["RPC call to ", queue, " couldn't be routed"] end
+  defp log_timeout(%{routing_key: routing_key} = request),
+    do: Logger.error fn -> ["RPC call to ", routing_key, " timed out after #{Request.duration(request)} ms"] end
```
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: Dynamic messages :/

Comment (deiwin) [inline lib/freddy/rpc/client.ex]: Is this sent to airbrake? We've been avoiding sending these things to airbrake, as their much more valuable when recorded as stats.

File: lib/freddy/rpc/client.ex [hunk 8]
```diff
@@ -488,19 +532,17 @@ defmodule Freddy.RPC.Client do
...
 
-  defp calculate_elapsed(%{start_time: start_time, stop_time: stop_time}),
-    do: System.convert_time_unit(stop_time - start_time, :native, :milliseconds)
-  defp calculate_elapsed(_),
-    do: :unknown
+  defp log_return(%{routing_key: routing_key} = _request),
+    do: Logger.error fn -> ["RPC call to ", routing_key, " couldn't be routed"] end
```
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: This can happen when the other service we're trying to talk to is down, right? This should probably also not make it to airbrake.

File: lib/freddy/rpc/request.ex [hunk 9]
```diff
@@ -0,0 +1,76 @@
+defmodule Freddy.RPC.Request do
+  @moduledoc """
+  RPC Request data structure. Can be modified by clients.
```
Comment (deiwin) [inline lib/freddy/rpc/request.ex]: Why would clients need to modify this?

Comment (deiwin) [inline lib/freddy/rpc/request.ex]: And do you mean the client of the library or the client in the RPC client<->server relationship?

File: test/freddy/rpc/client_test.exs [hunk 10]
```diff
@@ -39,51 +48,147 @@ defmodule Freddy.RPC.ClientTest do
...
+      modified_options = [app_id: "testapp2"]
+
+      options = [app_id: "testapp", hook: {:modify, %{
+        payload: modified_payload,
+        routing_key: modified_routing_key,
+        options: modified_options
+      }}]
+
+      request = Task.async fn ->
+        Freddy.RPC.Client.request(rpc_client, original_routing_key, original_payload, options)
```
Comment (deiwin) [inline test/freddy/rpc/client_test.exs]: I'm confused. Why does a single function call include 2 different routing keys, payloads and options?

**Comment by take-five** [comment]
> The example in the commit (and PR) message seems to omit the routing key altogether. I assume that that's not intentional, right?

Yes, mistake

File: lib/freddy/rpc/client.ex (hunk 3 above)
Comment (take-five) [inline lib/freddy/rpc/client.ex]: It really isnt't, but I've noticed that it'd be nice to react on returned message in application too (send stats to monitoring). It should've been in a separate commit probably

File: lib/freddy/rpc/client.ex (hunk 7 above)
Comment (take-five) [inline lib/freddy/rpc/client.ex]: > Is this sent to airbrake? We've been avoiding sending these things to airbrake, as their much more valuable when recorded as stats.

It's a log statement. We do not send errors to Airbrake based on logs, we monitor process exits instead

File: test/freddy/rpc/client_test.exs (hunk 10 above)
Comment (take-five) [inline test/freddy/rpc/client_test.exs]: > I'm confused. Why does a single function call include 2 different routing keys, payloads and options?

This tests checks behaviour of `before_request` callback, when it returns `{:ok, modified_request, new_state}` tuple.

File: lib/freddy/rpc/client.ex (hunk 4 above)
Comment (take-five) [inline lib/freddy/rpc/client.ex]: 5d0d35f

File: lib/freddy/rpc/client.ex (hunk 5 above)
Comment (take-five) [inline lib/freddy/rpc/client.ex]: 5d0d35f

File: lib/freddy/rpc/client.ex (hunk 6 above)
Comment (take-five) [inline lib/freddy/rpc/client.ex]: 5d0d35f

File: lib/freddy/rpc/client.ex (hunk 7 above)
Comment (take-five) [inline lib/freddy/rpc/client.ex]: What do you mean by dynamic messages btw?

**Comment by take-five** [comment]
Update commit message and PR description

**Comment by take-five** [comment]
I'll merge it now. If any of outstanding concerns are critical, I'll address them in a separate PRs

File: lib/freddy/rpc/client.ex (hunk 4 above)
Comment (take-five) [inline lib/freddy/rpc/client.ex]: > Why are we relying in exceptions instead of success/failure return values? Should send_request be changed to send_request! by the same convention?

Actually, now I think that I should've better leave exception here. If application sends something that can't be encoded, then it definitely does something wrong and we'd rather crash than try to handle this case. What do you think of it?

File: lib/freddy/rpc/client.ex (hunk 7 above)
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: > We do not send errors to Airbrake based on logs

Is that a problem? Because in every other project we send every `logger.error` to Airbrake.

Comment (deiwin) [inline lib/freddy/rpc/client.ex]: > What do you mean by dynamic messages btw?

The message has dynamic information in it. Such as the request duration here. Instead of the request duration being included in the metadata.

Comment (take-five) [inline lib/freddy/rpc/client.ex]: > Is that a problem? Because in every other project we send every logger.error to Airbrake.

Not true. Ruby projects send reports on exception, not because of logging statements with certain severity.

Anyway, let's not discuss it in public space

File: lib/freddy/rpc/client.ex (hunk 4 above)
Comment (deiwin) [inline lib/freddy/rpc/client.ex]: Well, we crash now, don't we? We send a `:stop` in that case. Not sure if stopping because of an exception would be better.

File: lib/freddy/rpc/client.ex (hunk 7 above)
Comment (take-five) [inline lib/freddy/rpc/client.ex]: > The message has dynamic information in it. Such as the request duration here. Instead of the request duration being included in the metadata.

This approach is specific to our company, whilst this library is OO. This is why I'm inclined to remove logging from Freddy completely and let application handle it by implementing callbacks
